### PR TITLE
feat(compass-aggregations): warn users when generated aggregation contains a write operation COMPASS-7298

### PR DESCRIPTION
**Pull request by alenakhineika**
## Description

Aggregation stages like `$merge` and `$out` can write to a customer's collection and alter their data. We should warn users before they execute it.

Screenshot 2024-04-11 at 19 49 14
Screenshot 2024-04-11 at 19 49 34
Screenshot 2024-04-11 at 19 49 44

For `S3` and other cases where we weren't able to extract a namespace from a stage:
Screenshot 2024-04-17 at 12 07 58

### Checklist
- [x] New tests and/or benchmarks are included
- [ ] Documentation is changed or added
- [ ] I have signed the MongoDB Contributor License Agreement (https://www.mongodb.com/legal/contributor-agreement)

## Motivation and Context

- [ ] Bugfix
- [x] New feature
- [ ] Dependency update
- [ ] Misc

## Types of changes

- [ ] *Backport Needed*
- [ ] Patch (non-breaking change which fixes an issue)
- [x] Minor (non-breaking change which adds functionality)
- [ ] Major (fix or feature that would cause existing functionality to change)

### CODE REVIEW

File: packages/compass-aggregations/src/modules/aggregation.ts [hunk 1]
```diff
@@ -204,9 +224,109 @@ const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
   return state;
 };
 
+const isConfirmReceivedOrNotRequired = async (
```
Comment (mcasimir) [inline packages/compass-aggregations/src/modules/aggregation.ts:227]: nit: what about calling this just `confirmWriteOperation`? I was thinking that for someone without context, knowing that this is for write operations at call point can be valuable piece of info. I would be able to infer that is ignoring non "write" pipelines. Otherwise we can also say `confirmWriteOperationIfNeeded`

File: packages/compass-aggregations/src/modules/aggregation.ts [hunk 2]
```diff
@@ -204,9 +224,109 @@ const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
...
+  let { database, collection } = toNS(namespace);
+
+  if (typeof lastStage.$merge === 'string') {
+    collection = lastStage.$merge;
+  }
+
+  if (
+    typeof lastStage.$merge === 'object' &&
+    lastStage.$merge !== null &&
+    typeof lastStage.$merge.into === 'string'
```
Comment (mcasimir) [inline packages/compass-aggregations/src/modules/aggregation.ts:250]: do we want to handle also ADF here?

https://www.mongodb.com/docs/atlas/data-federation/supported-unsupported/pipeline/merge/

Comment (gribnoysup) [reply packages/compass-aggregations/src/modules/aggregation.ts:250]: We have a bunch of helpers in agg plugin to deal with this (getting namepace, checking if out stage, etc) https://github.com/mongodb-js/compass/blob/a41dd2de57e01ec0ef79f8697be86dc36a299c37/packages/compass-aggregations/src/utils/stage.ts#L111-L132

File: packages/compass-e2e-tests/tests/collection-aggregations-tab.test.ts [hunk 3]
```diff
@@ -688,6 +862,198 @@ describe('Collection aggregations tab', function () {
     );
   });
 
+  it('finds $merge destination as { $merge: <collection> }', async function () {
```
Comment (mcasimir) [inline packages/compass-e2e-tests/tests/collection-aggregations-tab.test.ts:865]: Nit: are we sure that we want to test all of the conditions as e2e tests? if we want to cover all of the corner cases and possible variations it may need a good number of them 

how about we extract a `getWriteStageInfo(serverType, ns, pipeline, ...) -> {writeStage, destinationNamespace} | undefined` function, which is a pure function, and we write 200 unit tests for that, covering all sort of corner cases. Here we can just have one e2e test to make sure  the confirmation gets shown on the most simple case.

File: packages/compass-aggregations/src/modules/aggregation.ts (hunk 1 above)
Comment (alenakhineika) [reply packages/compass-aggregations/src/modules/aggregation.ts:227]: I will update with `confirmWriteOperationIfNeeded` 👍

File: packages/compass-aggregations/src/modules/aggregation.ts (hunk 2 above)
Comment (alenakhineika) [reply packages/compass-aggregations/src/modules/aggregation.ts:250]: @gribnoysup nice 👍 will refactor with using those!

File: packages/compass-e2e-tests/tests/collection-aggregations-tab.test.ts (hunk 3 above)
Comment (alenakhineika) [reply packages/compass-e2e-tests/tests/collection-aggregations-tab.test.ts:865]: It is not needed since Sergey pointed to existing helpers to extract destinations from stages. Will refactor accordingly and remove redundant tests. Please note that this is a draft PR.

Comment (mcasimir) [reply packages/compass-e2e-tests/tests/collection-aggregations-tab.test.ts:865]: Sure, no worries!

File: packages/compass-aggregations/src/modules/aggregation.ts [hunk 4]
```diff
@@ -204,9 +224,109 @@ const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
...
+) => {
+  const lastStage = pipeline[pipeline.length - 1];
+  const lastStageName = Object.keys(lastStage)[0];
+  let typeOfWrite;
+
+  if (!['$merge', '$out'].includes(lastStageName)) {
+    return true;
+  }
+
+  let { database, collection } = toNS(namespace);
```
[the remark this answers is not in the document]
Comment (alenakhineika) [reply packages/compass-aggregations/src/modules/aggregation.ts:241]: According to syntax you may or may not specify the database:

https://www.mongodb.com/docs/manual/reference/operator/aggregation/out/
https://www.mongodb.com/docs/manual/reference/operator/aggregation/merge/

If db is not specified then the destination is the correct database. As discussed in another thread here, I will refactor the code with the usage of existing helpers that extract the destination namespace from the stage. And will check if this makes sense to show some kind of validation if this fails.

File: packages/compass-aggregations/src/modules/aggregation.ts [hunk 5]
```diff
@@ -204,9 +225,89 @@ const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
...
+  const lastStageOperator = getStageOperator(pipeline[pipeline.length - 1]);
+  let typeOfWrite;
+
+  if (!lastStageOperator || !isOutputStage(lastStageOperator)) {
+    return true;
+  }
+
+  const destinationNamespace = getDestinationNamespaceFromStage(namespace, {
+    [lastStageOperator]: pipeline[pipeline.length - 1][lastStageOperator],
+  });
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/modules/aggregation.ts:247]: This is just this, right?

```suggestion
  const destinationNamespace = getDestinationNamespaceFromStage(namespace, pipeline[pipeline.length - 1]);
```

File: packages/compass-aggregations/src/modules/aggregation.ts [hunk 6]
```diff
@@ -204,9 +225,89 @@ const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
...
+  instance: MongoDBInstance,
+  dataService: DataService,
+  namespace: string,
+  pipeline: Document[]
+) => {
+  if (!pipeline.length) {
+    return true;
+  }
+
+  const lastStageOperator = getStageOperator(pipeline[pipeline.length - 1]);
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/modules/aggregation.ts:238]: Just FYI, `getStageOperator` supports nullish values, so you can just do one check for `lastStageOperator` below instead of checking the array length first
```suggestion
  const lastStageOperator = getStageOperator(pipeline[pipeline.length - 1]);
```

File: packages/compass-aggregations/src/modules/aggregation.ts [hunk 7]
```diff
@@ -204,9 +225,89 @@ const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
...
+  if (!lastStageOperator || !isOutputStage(lastStageOperator)) {
+    return true;
+  }
+
+  const destinationNamespace = getDestinationNamespaceFromStage(namespace, {
+    [lastStageOperator]: pipeline[pipeline.length - 1][lastStageOperator],
+  });
+
+  if (lastStageOperator === '$out') {
+    let isOverwritingCollection;
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/modules/aggregation.ts:250]: To avoid this being of type any, you probably want to indicate what type you expect to assign here
```suggestion
    let isOverwritingCollection: boolean;
```

File: packages/compass-aggregations/src/modules/aggregation.ts [hunk 8]
```diff
@@ -204,9 +225,89 @@ const reducer: Reducer<State> = (state = INITIAL_STATE, action) => {
...
+    if (
+      !(await confirmWriteOperationIfNeeded(
+        instance,
+        dataService,
+        state.namespace,
+        pipeline
+      ))
+    ) {
+      return;
+    }
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/modules/aggregation.ts:309]: It's an antipattern to store state in a variable inside thunk actions. Even though both calls here are sync, let's keep the pattern just so it's not confusing that we do one thing here, but a different one elsewhere in this code:

```suggestion
    const pipeline = getPipelineFromBuilderState(getState(), pipelineBuilder);

    if (
      !(await confirmWriteOperationIfNeeded(
        instance,
        dataService,
        getState().namespace,
        pipeline
      ))
    ) {
      return;
    }
```

File: packages/compass-aggregations/src/utils/modal-descriptions.tsx [hunk 9]
```diff
@@ -0,0 +1,35 @@
...
+export const runPipelineConfirmationDescription = ({
+  typeOfWrite,
+  stage,
+  ns,
+}: {
+  typeOfWrite: string;
+  stage: { name: string; link: string };
+  ns: string | null;
+}) => {
+  if (!ns) {
```
Comment (gribnoysup) [inline packages/compass-aggregations/src/utils/modal-descriptions.tsx:13]: Nit: seems like the only difference is the text after the comma, would probably be easier to update shared parts if it's all the same and only what's different is changing:

```
const nsText = ns ? <>{typeOfWrite} &quot;<b>{ns}</b>&quot;<> : <>that may alter or overwrite a collection<>

// ...

operation, {nsText}. Do you wish to proceed?
```

File: packages/compass-e2e-tests/helpers/selectors.ts [hunk 10]
```diff
@@ -714,6 +714,11 @@ export const AggregationRestultsPrevPageButton =
 export const AggregationResultsCancelButton =
   '[data-testid="pipeline-results-loader-button"]';
 export const AggregationEmptyResults = '[data-testid="pipeline-empty-results"]';
+export const AggregationWriteOperationConfirmationModal =
+  '[data-testid="write-operation-confirmation-modal"]';
+export const AggregationWriteOperationConfirmButton = `${AggregationWriteOperationConfirmationModal} button:first-of-type`;
+export const AggregationWriteOperationCancelButton = `${AggregationWriteOperationConfirmationModal} button:last-of-type`;
```
Comment (gribnoysup) [inline packages/compass-e2e-tests/helpers/selectors.ts:720]: There are testids on those elements that we can use instead:

```suggestion
export const AggregationWriteOperationConfirmButton = `${AggregationWriteOperationConfirmationModal} button:[data-testid*="confirm-button"]`;
export const AggregationWriteOperationCancelButton = `${AggregationWriteOperationConfirmationModal} [data-testid*="cancel-button"]`;
```

File: packages/compass-e2e-tests/helpers/selectors.ts [hunk 11]
```diff
@@ -714,6 +714,11 @@ export const AggregationRestultsPrevPageButton =
...
+export const AggregationWriteOperationConfirmationModalDescription = `${AggregationWriteOperationConfirmationModal} > div > div`;
```
Comment (gribnoysup) [inline packages/compass-e2e-tests/helpers/selectors.ts:721]: Let's add a wrapper with testid instead of targeting nested div if we really need to get this particular element selected

**Review by gribnoysup — APPROVED** [review]
Some small comments here and there, but overall looks good

File: packages/compass-aggregations/src/modules/aggregation.ts (hunk 5 above)
Comment (alenakhineika) [reply packages/compass-aggregations/src/modules/aggregation.ts:247]: Lol yeah 🙈
